### DISCUSSION TITLE
How to detect app activation on macOS

### DISCUSSION THREAD

**Discussion by pawelmajda** [AvaloniaUI/Avalonia#7688] [Q&A]
**Question**

How to detect event of app being activated (when selecting it from the dock) after it was previously hidden in macOS?

**Full Context**

I am working on a cross-platform desktop app and on macOS, I am trying to achieve the behaviour of closing/hiding just the UI (and not the whole app) when the main window is closed (using system close button). And then when the app is activated again (by selecting it from the dock), reopening and showing the main window back. That is the usual behaviour on macOS for any app that is expected to work in the background.

I was able to achieve the first part by using `IClassicDesktopStyleApplicationLifetime` with `OnExplicitShutdown` shutdown mode (as described in https://github.com/AvaloniaUI/Avalonia/wiki/Application-lifetimes#iclassicdesktopstyleapplicationlifetime) or by hiding the main window instead of closing (as described in https://docs.avaloniaui.net/docs/controls/window#prevent-a-window-from-closing). 

Nevertheless, I didn't find any information in docs regarding how to detect that the app was reactivated or brought to the foreground again, so I can reopen and show the main window back in that moment. Is there any API for that exposed by Avalonia? I only found native macOS API for that (like https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428585-applicationwillunhide) but can it be accessed via Avalonia? Or maybe there is some other approach to achieve such behaviour?

**Comment by maxkatz6** [comment]
I can't answer your question specifically now. Maybe somebody else can.

But you always can access native APIs from Avalonia app.
There are always features that aren't or can't be covered by GUI framework.
To use macOS APIs you can either target "net6-macos" or use MonoMac dependency. Either way you will be able to use APIs like [NSApplication](https://developer.apple.com/documentation/appkit/nsapplication) and more, which can be used to listen for native events or modify app configuration directly.

**Comment by mjrj97** [comment]
Did you find a solution to the problem @pawelmajda? I'm also trying to solve the same problem.

**Reply by mjrj97** [reply]
I managed to find a solution. Using the Avalonia.MonoMac nuget package, I the first two lines in the Initialize method of my App class, and OpenMainWindow is just a simple method. I read somewhere that the MonoMac package isn't being supported anymore, but this is the best I could find. I guess you could just copy the approach the package uses.
```
NSApplication.Init();
NSApplication.Notifications.ObserveDidBecomeActive(OpenMainWindow);

---------------

public void OpenMainWindow(object sender, NSNotificationEventArgs e)
{
    Dispatcher.UIThread.InvokeAsync(delegate
    {
        window.Show();
    });
}
```
